src/static_allocator/device.rs

```diff
@@ -27,16 +24,71 @@ pub struct StaticDeviceAllocator {
     // TODO: Can we use deque
     bitmap: Arc<Vec<AtomicBool>>,
     #[cfg(feature = "allocator_stats")]
-    maximum_tail_index: Arc<AtomicUsize>,
+    pub stats: Arc<std::sync::Mutex<AllocationStats>>,
 }
 
 #[cfg(feature = "allocator_stats")]
 #[derive(Derivative)]
-#[derivative(Clone, Debug)]
+#[derivative(Clone, Debug, Default)]
 pub struct AllocationStats {
-    pub current_block_count: usize,
-    pub current_tail_index: usize,
+    pub maximum_block_count: usize,
     pub maximum_tail_index: usize,
+    pub allocations: std::collections::BTreeMap<usize, (usize, String)>,
+}
+
+#[cfg(feature = "allocator_stats")]
+impl AllocationStats {
+    pub fn alloc(&mut self, index: usize, size: usize, backtrace: String) {
+        self.allocations.insert(index, (size, backtrace));
+        self.maximum_block_count = self.maximum_block_count.max(self.current_block_count());
+        self.maximum_tail_index = self.maximum_tail_index.max(index + size);
```

Comment:
In practice, I think we sometimes want to know which part of the code incurred the block count and tail index high water marks.
Can you add two fields
```rust
backtrace_at_maximum_block_count: String;
backtrace_at_maximum_tail_index: String;
```
and also update these whenever their corresponding maximums get updated? (and print them of course, your choice if detailed or not detailed)